# iNaturalist status updates by state

Using the file produced in the collate-status-taxa.ipynb: `inat-aust-status-taxa.csv`, generate lists to update iNaturalist statuses

**Next steps:**
State by state establish the changes that need to be made:
    a. new - any new species that appear in the state lists but do not have a status in inaturalist (new template)
    b. updates - any changes to information which was added by us previously (user_id = 708886) (update template, action='UPDATE')
    c. removals - any statuses which were added by us previously (user_id = 708886) list which are incorrect (update template, action='REMOVE')
    d. flags - are there any statuses by other users that need to be flagged?

## Prep - common to all states
1. Read in the inaturalist statuses & filter out this state
2. Read in the inaturalist taxa list
3. Read in the state sensitive list
4. Attempt to match the state statuses to an IUCN equivalent


### 1. iNaturalist statuses

In [27]:
import sys
import os
projectdir = "/Users/oco115/PycharmProjects/authoritative-lists/" # basedir for this gh project
sys.path.append(os.path.abspath(projectdir + "source-code/includes"))
import list_functions as lf

import pandas as pd
# projectdir = "/Users/new330/IdeaProjects/authoritative-lists/" # basedir for this gh project
sourcedir = projectdir + "source-data/inaturalist-statuses/"
listdir = projectdir + "current-lists/"


# read in the statuses
taxastatus = pd.read_csv(sourcedir + "inat-aust-status-taxa.csv", encoding='UTF-8',na_filter=False,dtype=str) ## Read inaturalist conservation statuses file
taxastatus.head(3)

id taxon_id  user_id place_id source_id  \
0  166449    38493  1138587     7830             
1  234788   918383   702203     9994             
2  234789   918383   702203     7308             

                            authority status  \
0  Flora and Fauna Guarantee Act 1988     CR   
1           Atlas of Living Australia    NT    
2           Atlas of Living Australia     LC   

                                                 url description geoprivacy  \
0                                                                  obscured   
1  https://bie.ala.org.au/species/https://id.biod...                          
2  https://bie.ala.org.au/species/https://id.biod...                          

   ...         genus specificEpithet infraspecificEpithet  \
0  ...     Eulamprus      kosciuskoi                        
1  ...  Chiloschista     phyllorhiza                        
2  ...  Chiloschista     phyllorhiza                        

               modified            scientificName taxonRank  \
0  2021-03-01T10:35:01Z      Eulamprus kosciuskoi   species   
1  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   
2  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   

                                          references preferred_common_name  \
0  http://reptile-database.reptarium.cz/search.ph...                         
1  http://www.catalogueoflife.org/annual-checklis...                         
2  http://www.catalogueoflife.org/annual-checklis...                         

  is_active current_synonymous_taxon_ids  
0                                         
1                                         
2                                         

[3 rows x 34 columns]

In [28]:
def filter_state_statuses(stateregex: str, urlregex: str):
    authoritydf = taxastatus['authority'].drop_duplicates().sort_values()
    authoritydf = authoritydf[pd.Series(authoritydf).str.contains(stateregex)]
    urldf = taxastatus['url'].drop_duplicates().sort_values()
    urldf = urldf[pd.Series(urldf).str.contains(urlregex)]
    placedisplaydf = taxastatus['place_display_name'].drop_duplicates().sort_values()
    placedisplaydf = placedisplaydf[pd.Series(placedisplaydf).str.contains(stateregex)]
    placedf = taxastatus['place_name'].drop_duplicates().sort_values()
    placedf = placedf[pd.Series(placedf).str.contains(stateregex)]
    # concat all and remove duplicates
    statedf = pd.concat([taxastatus.apply(lambda row: row[taxastatus['place_display_name'].isin(placedisplaydf)]),
                         taxastatus.apply(lambda row: row[taxastatus['place_name'].isin(placedf)]),
                         taxastatus.apply(lambda row: row[taxastatus['url'].isin(urldf)]),
                         taxastatus.apply(
                             lambda row: row[taxastatus['authority'].isin(authoritydf)])]).drop_duplicates()
    return statedf.sort_values(['taxon_id', 'user_id'])
inatstatuses = filter_state_statuses("Tasmania|TAS", ".tas.gov")
inatstatuses.rename(columns={'id':'status_id','id_y':'taxon_id_y'},inplace=True)
inatstatuses

status_id taxon_id  user_id place_id source_id  \
212     170159  1007040   702203     6829             
2558    160921  1038747   717122     6829             
2547    160910  1038906   717122     6829             
933     160738  1038964   717122     6829             
2557    160920  1038976   717122     6829             
...        ...      ...      ...      ...       ...   
2586    160759   958863   717122     6829             
2594    161103   960004   717122     6829             
3380    268029   993886  1116633     6829             
813     160748   993902   717122     6829             
1008    160745   993922   717122     6829             

                                            authority      status  \
212            Threatened Species Protection Act 1995          EN   
2558           Threatened Species Protection Act 1995  Endangered   
2547           Threatened Species Protection Act 1995  Endangered   
933            Threatened Species Protection Act 1995  Endangered   
2557           Threatened Species Protection Act 1995  Endangered   
...                                               ...         ...   
2586           Threatened Species Protection Act 1995  Endangered   
2594           Threatened Species Protection Act 1995  Endangered   
3380  TASMANIAN THREATENED SPECIES LISTING STATEMENT           EN   
813            Threatened Species Protection Act 1995  Endangered   
1008           Threatened Species Protection Act 1995  Endangered   

                                                    url  \
212   https://www.threatenedspecieslink.tas.gov.au/P...   
2558  https://dpipwe.tas.gov.au/Documents/Ranunculus...   
2547  https://dpipwe.tas.gov.au/Documents/Myosurus-a...   
933   https://dpipwe.tas.gov.au/Documents/Boronia%20...   
2557  https://dpipwe.tas.gov.au/Documents/Pimelea-ax...   
...                                                 ...   
2586  https://dpipwe.tas.gov.au/Documents/Corunastyl...   
2594  https://dpipwe.tas.gov.au/Documents/Thelymitra...   
3380  https://nre.tas.gov.au/Documents/Prasophyllum%...   
813   https://dpipwe.tas.gov.au/Documents/Caladenia-...   
1008  https://dpipwe.tas.gov.au/Documents/Caladenia-...   

                                            description geoprivacy  ...  \
212                                                                 ...   
2558                                                      obscured  ...   
2547                                                      obscured  ...   
933                                                       obscured  ...   
2557                                                      obscured  ...   
...                                                 ...        ...  ...   
2586                                                                ...   
2594                                                      obscured  ...   
3380  1. total population estimated to number fewer ...   obscured  ...   
813                                                                 ...   
1008                                                                ...   

             genus specificEpithet infraspecificEpithet              modified  \
212       Dianella          amoena                       2021-07-27T17:12:17Z   
2558    Ranunculus       diminutus                       2022-06-13T17:38:59Z   
2547      Myosurus       australis                       2020-08-19T03:33:32Z   
933        Boronia      hemichiton                       2020-08-17T05:51:50Z   
2557       Pimelea        axiflora             axiflora  2020-12-02T04:00:26Z   
...            ...             ...                  ...                   ...   
2586   Genoplesium      nudiscapum                       2022-10-26T09:34:12Z   
2594    Thelymitra       bracteata                       2022-07-14T00:09:44Z   
3380  Prasophyllum    abblittiorum                       2020-02-18T23:45:10Z   
813      Caladenia         dienema                       2020-12-15T00:50:01Z   
1008     Calad

### 2. iNaturalist taxonomy

In [29]:
# Output files contain these fields
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# so we need to match species from the state lists to the inat taxa to get the taxon_id

import zipfile
url = "https://www.inaturalist.org/taxa/inaturalist-taxonomy.dwca.zip"
filename = url.split("/")[-1]

z=zipfile.ZipFile(sourcedir + filename)

with z.open('taxa.csv') as from_archive:
    inattaxa = pd.read_csv(from_archive,dtype=str)
z.close()
inattaxa.head(3)


id                             taxonID                          identifier  \
0  1  https://www.inaturalist.org/taxa/1  https://www.inaturalist.org/taxa/1   
1  2  https://www.inaturalist.org/taxa/2  https://www.inaturalist.org/taxa/2   
2  3  https://www.inaturalist.org/taxa/3  https://www.inaturalist.org/taxa/3   

                         parentNameUsageID   kingdom    phylum class order  \
0   https://www.inaturalist.org/taxa/48460  Animalia       NaN   NaN   NaN   
1       https://www.inaturalist.org/taxa/1  Animalia  Chordata   NaN   NaN   
2  https://www.inaturalist.org/taxa/355675  Animalia  Chordata  Aves   NaN   

  family genus specificEpithet infraspecificEpithet              modified  \
0    NaN   NaN             NaN                  NaN  2021-11-02T06:05:44Z   
1    NaN   NaN             NaN                  NaN  2021-11-23T00:40:18Z   
2    NaN   NaN             NaN                  NaN  2022-12-27T07:33:16Z   

  scientificName taxonRank                                         references  
0       Animalia   kingdom  http://www.catalogueoflife.org/annual-checklis...  
1       Chordata    phylum  http://www.catalogueoflife.org/annual-checklis...  
2           Aves     class  http://www.catalogueoflife.org/annual-checklis...

### 3. State lists
Sensitive list: `geoprivacy` = `obscured`
overrides

In [30]:

# %%script echo skipping # comment this line to download dataset from lists.ala.org.au the web and save locally
# Download lists data. Retrieve binomial and trinomial names from GBIF. Save locally to CSV

sensitivelist = lf.download_ala_list("https://lists.ala.org.au/ws/speciesListItems/dr491?max=10000&includeKVP=true")
# sensitivelist = lf.kvp_to_columns(sensitivelist)   # kvp values null
sensitivelist.to_csv(sourcedir + "tas-ala-sensitive.csv", index=False)

conservationlist = lf.download_ala_list("https://lists.ala.org.au/ws/speciesListItems/dr654?max=10000&includeKVP=true")
conservationlist = lf.kvp_to_columns(conservationlist)
conservationlist.to_csv(sourcedir + "tas-ala-conservation.csv", index=False)

In [31]:
# Read sensitive list data
sensitivelist = pd.read_csv(sourcedir + "tas-ala-sensitive.csv", dtype=str)
sensitivelist['status'] = 'sensitive'
sensitivelist['bionet_geoprivacy'] = 'obscured'
sensitivelist

id               name      commonName     scientificName  \
0  24386  Lomatia tasmanica  King's Lomatia  Lomatia tasmanica   

                                               lsid dataResourceUid kvpValues  \
0  https://id.biodiversity.org.au/node/apni/2903191           dr491        []   

      status bionet_geoprivacy  
0  sensitive          obscured

In [32]:
# Read conservation list data
conservationlist = pd.read_csv(sourcedir + "tas-ala-conservation.csv", dtype=str)
conservationlist['bionet_geoprivacy'] = 'open'
conservationlist

id                                    name  \
0    4033150                           Costora iena    
1    4033128                  Hadronyche pulvinator    
2    4033378                 Hypolimnus pedderensis    
3    4033234                Thylacinus cynocephalus    
4    4033416  Gallirallus philippensis macquariensis   
..       ...                                     ...   
678  4033369                    Erioderma sorediatum   
679  4033419              Xanthoparmelia amphixantha   
680  4033291              Xanthoparmelia molliuscula   
681  4033236             Xanthoparmelia subloxodella   
682  4033313                 Xanthoparmelia willisii   

                    commonName                          scientificName  \
0     Caddis Fly (great Lakes)                            Costora iena   
1    Cascade Funnel-web Spider                   Hadronyche pulvinator   
2        Lake Pedder Earthworm                  Hypolimnus pedderensis   
3                    Thylacine                 Thylacinus cynocephalus   
4                          NaN  Gallirallus philippensis macquariensis   
..                         ...                                     ...   
678                        NaN                    Erioderma sorediatum   
679                        NaN              Xanthoparmelia amphixantha   
680                        NaN              Xanthoparmelia molliuscula   
681                        NaN             Xanthoparmelia subloxodella   
682                        NaN                 Xanthoparmelia willisii   

                                                  lsid dataResourceUid  \
0    https://biodiversity.org.au/afd/taxa/7de808fb-...           dr654   
1    https://biodiversity.org.au/afd/taxa/2acd0df1-...           dr654   
2    https://biodiversity.org.au/afd/taxa/7db09aa3-...           dr654   
3    https://biodiversity.org.au/afd/taxa/97764bed-...           dr654   
4                                        NZOR-6-101588           dr654   
..                                                 ...             ...   
678                                      ALA_DR654_679           dr654   
679                                      ALA_DR654_680           dr654   
680                                      ALA_DR654_681           dr654   
681                                      ALA_DR654_682           dr654   
682                                      ALA_DR654_683           dr654   

                                             kvpValues  \
0    [{'key': 'scientificNameAuthorship', 'value': ...   
1    [{'key': 'scientificNameAuthorship', 'value': ...   
2    [{'key': 'scientificNameAuthorship', 'value': ...   
3    [{'key': 'scientificNameAuthorship', 'value': ...   
4    [{'key': 'scientificNameAuthorship', 'value': ...   
..                                                 ...   
678  [{'key': 'scientificNameAuthorship', 'value': ...   
679  [{'key': 'scientificNameAuthorship', 'value': ...   
680  [{'key': 'scientificNameAuthorship', 'value': ...   
681  [{'key': 'scientificNameAuthorship', 'value': ...   
682  [{'key': 'scientificNameAuthorship', 'value': ...   

                              scientificNameAuthorship  \
0                                               Mosely   
1                                              Hickman   
2                                           (Jamieson)   
3                                               Harris   
4                                               Hutton   
..                                                 ...   
678                           D.J.Galloway & P.M.Jørg.   
679                                   (Müll.Arg.) Hale   
680                                        (Ach.) Hale   
681  (Elix & Kantvilas) O.Blanco, A.Crespo, Elix, D...   
682                 (Kurok. & Filson) Elix & J.Johnst.   

               vernacular name        family          class sourceStatus  \
0     Caddis Fly (Great Lakes)   Trichoptera  Invertebrates            x   
1    Cascade 

In [33]:
# join them in a way that works for inat (eg sensitive list, geoprivacy = 'obscured'
statelist = pd.concat([sensitivelist[['name','bionet_geoprivacy', 'status', 'lsid']],
                    conservationlist[['name','bionet_geoprivacy', 'status', 'lsid']]]).drop_duplicates()
statelist

name bionet_geoprivacy            status  \
0              Lomatia tasmanica          obscured         sensitive   
0                  Costora iena               open  Presumed Extinct   
1         Hadronyche pulvinator               open  Presumed Extinct   
2        Hypolimnus pedderensis               open  Presumed Extinct   
3       Thylacinus cynocephalus               open  Presumed Extinct   
..                           ...               ...               ...   
678         Erioderma sorediatum              open        Endangered   
679   Xanthoparmelia amphixantha              open        Endangered   
680   Xanthoparmelia molliuscula              open        Endangered   
681  Xanthoparmelia subloxodella              open        Endangered   
682      Xanthoparmelia willisii              open        Endangered   

                                                  lsid  
0     https://id.biodiversity.org.au/node/apni/2903191  
0    https://biodiversity.org.au/afd/taxa/7de808fb-...  
1    https://biodiversity.org.au/afd/taxa/2acd0df1-...  
2    https://biodiversity.org.au/afd/taxa/7db09aa3-...  
3    https://biodiversity.org.au/afd/taxa/97764bed-...  
..                                                 ...  
678                                      ALA_DR654_679  
679                                      ALA_DR654_680  
680                                      ALA_DR654_681  
681                                      ALA_DR654_682  
682                                      ALA_DR654_683  

[684 rows x 4 columns]

In [34]:
# check for duplicates with conflicting information
statelist.groupby('name').filter(lambda x: len(x) > 1)#.sort('size',ascending=False)


name bionet_geoprivacy      status  \
0    Lomatia tasmanica          obscured   sensitive   
569  Lomatia tasmanica              open  Endangered   

                                                 lsid  
0    https://id.biodiversity.org.au/node/apni/2903191  
569  https://id.biodiversity.org.au/node/apni/2903191

### 4. Equivalent IUCN statuses

In [35]:
# iucn_statuses = {'Not Evaluated', 'Data Deficient', 'Least Concern', 'Near Threatened', 'Vulnerable', 'Endangered', 'Critically Endangered', 'Extinct in the Wild', 'Extinct'}
statelist.groupby(['status'])['status'].count()

status
Endangered          222
Presumed Extinct     28
Rare                311
Vulnerable          122
sensitive             1
Name: status, dtype: int64

In [36]:
iucnStatusMappings = {
    'Presumed Extinct':'Extinct',
    'Rare':'Vulnerable',
    'sensitive':'Vulnerable'
}

### 5. Determine best place ID to use

In [37]:
inatstatuses.groupby(['place_id','place_name','place_display_name'])['place_id'].count()
# looks like 6825

place_id  place_name          place_display_name    
6744      Australia           Australia                  2
6829      Tasmania            Tasmania, AU              77
9994      Northern Territory  Northern Territory, AU     6
Name: place_id, dtype: int64

## Merge iNaturalist statuses with State sensitive list on scientificName

1. Match - updates, even if the statuses are the same we'll update the links and values anyway
2. No match - statuses to be added (additions)
   1.1 No match and no taxnomy - search for synonyms
   1.2 No match
3. Merge the other direction to see if there are deletes?


In [38]:
# join to see which lists already have a status in inaturalist based on scientificName
mergedstatuses = statelist[['name', 'bionet_geoprivacy', 'lsid']].merge(inatstatuses[['status_id','scientificName','taxon_id','user_id','description','iucn','authority','status','geoprivacy','place_id','place_display_name']],how="left",left_on='name',right_on='scientificName',suffixes=(None,'_inat')).sort_values(['name'])
mergedstatuses


name bionet_geoprivacy  \
216                   Acacia axillaris              open   
300                  Acacia pataczekii              open   
355                Acacia siculiformis              open   
356                  Acacia ulicifolia              open   
357                   Acacia uncifolia              open   
..                                 ...               ...   
287               Xerochrysum palustre              open   
173                   Zearaja maugeana              open   
522                  Zieria littoralis              open   
678  Zieria veronicea subsp. veronicea              open   
523           Zygophyllum billardierei              open   

                                                  lsid status_id  \
216   https://id.biodiversity.org.au/node/apni/2911902    160726   
300   https://id.biodiversity.org.au/node/apni/2913609       NaN   
355   https://id.biodiversity.org.au/node/apni/2888960       NaN   
356   https://id.biodiversity.org.au/node/apni/2907350       NaN   
357   https://id.biodiversity.org.au/node/apni/2916270       NaN   
..                                                 ...       ...   
287   https://id.biodiversity.org.au/node/apni/2890165       NaN   
173  https://biodiversity.org.au/afd/taxa/0d9d194f-...       NaN   
522   https://id.biodiversity.org.au/node/apni/2888662       NaN   
678   https://id.biodiversity.org.au/node/apni/2911634       NaN   
523  https://id.biodiversity.org.au/taxon/apni/5131...       NaN   

       scientificName taxon_id user_id description iucn  \
216  Acacia axillaris   775192  717122               30   
300               NaN      NaN     NaN         NaN  NaN   
355               NaN      NaN     NaN         NaN  NaN   
356               NaN      NaN     NaN         NaN  NaN   
357               NaN      NaN     NaN         NaN  NaN   
..                ...      ...     ...         ...  ...   
287               NaN      NaN     NaN         NaN  NaN   
173               NaN      NaN     NaN         NaN  NaN   
522               NaN      NaN     NaN         NaN  NaN   
678               NaN      NaN     NaN         NaN  NaN   
523               NaN      NaN     NaN         NaN  NaN   

                                             authority status geoprivacy  \
216  Environment Protection and Biodiversity Conser...     VU   obscured   
300                                                NaN    NaN        NaN   
355                                                NaN    NaN        NaN   
356                                                NaN    NaN        NaN   
357                                                NaN    NaN        NaN   
..                                                 ...    ...        ...   
287                                                NaN    NaN        NaN   
173                                                NaN    NaN        NaN   
522                                                NaN    NaN        NaN   
678                                                NaN    NaN        NaN   
523                                                NaN    NaN        NaN   

    place_id place_display_name  
216     6744          Australia  
300      NaN                NaN  
355      NaN                NaN  
356      NaN                NaN  
357      NaN                NaN  
..       ...                ...  
287      NaN                NaN  
173      NaN                NaN  
522      NaN                NaN  
678      NaN                NaN  
523      NaN                NaN  

[684 rows x 14 columns]

In [39]:
# prepare the export fields, common to New template and Update template
# new statuses
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username
mergedstatuses['new_authority'] = "Threatened Species Protection Act 1995"
mergedstatuses['new_description'] = "Listed as Threatened - refer to https://www.threatenedspecieslink.tas.gov.au"
mergedstatuses['new_url'] = mergedstatuses['lsid']
mergedstatuses['new_geoprivacy'] = "obscured"
mergedstatuses['new_place_id'] = '6829'  # Tasmania
mergedstatuses['new_username'] = 'peggydnew'
mergedstatuses['new_iucn_equivalent'] = mergedstatuses['status'].str.lower().str.strip().map(iucnStatusMappings).fillna('Vulnerable') # map to dictionary
mergedstatuses['new_status'] = mergedstatuses['status'].fillna('Threatened')
mergedstatuses

name bionet_geoprivacy  \
216                   Acacia axillaris              open   
300                  Acacia pataczekii              open   
355                Acacia siculiformis              open   
356                  Acacia ulicifolia              open   
357                   Acacia uncifolia              open   
..                                 ...               ...   
287               Xerochrysum palustre              open   
173                   Zearaja maugeana              open   
522                  Zieria littoralis              open   
678  Zieria veronicea subsp. veronicea              open   
523           Zygophyllum billardierei              open   

                                                  lsid status_id  \
216   https://id.biodiversity.org.au/node/apni/2911902    160726   
300   https://id.biodiversity.org.au/node/apni/2913609       NaN   
355   https://id.biodiversity.org.au/node/apni/2888960       NaN   
356   https://id.biodiversity.org.au/node/apni/2907350       NaN   
357   https://id.biodiversity.org.au/node/apni/2916270       NaN   
..                                                 ...       ...   
287   https://id.biodiversity.org.au/node/apni/2890165       NaN   
173  https://biodiversity.org.au/afd/taxa/0d9d194f-...       NaN   
522   https://id.biodiversity.org.au/node/apni/2888662       NaN   
678   https://id.biodiversity.org.au/node/apni/2911634       NaN   
523  https://id.biodiversity.org.au/taxon/apni/5131...       NaN   

       scientificName taxon_id user_id description iucn  \
216  Acacia axillaris   775192  717122               30   
300               NaN      NaN     NaN         NaN  NaN   
355               NaN      NaN     NaN         NaN  NaN   
356               NaN      NaN     NaN         NaN  NaN   
357               NaN      NaN     NaN         NaN  NaN   
..                ...      ...     ...         ...  ...   
287               NaN      NaN     NaN         NaN  NaN   
173               NaN      NaN     NaN         NaN  NaN   
522               NaN      NaN     NaN         NaN  NaN   
678               NaN      NaN     NaN         NaN  NaN   
523               NaN      NaN     NaN         NaN  NaN   

                                             authority  ... place_id  \
216  Environment Protection and Biodiversity Conser...  ...     6744   
300                                                NaN  ...      NaN   
355                                                NaN  ...      NaN   
356                                                NaN  ...      NaN   
357                                                NaN  ...      NaN   
..                                                 ...  ...      ...   
287                                                NaN  ...      NaN   
173                                                NaN  ...      NaN   
522                                                NaN  ...      NaN   
678                                                NaN  ...      NaN   
523                                                NaN  ...      NaN   

    place_display_name                           new_authority  \
216          Australia  Threatened Species Protection Act 1995   
300                NaN  Threatened Species Protection Act 1995   
355                NaN  Threatened Species Protection Act 1995   
356                NaN  Threatened Species Protection Act 1995   
357                NaN  Threatened Species Protection Act 1995   
..                 ...                                     ...   
287                NaN  Threatened Species Protection Act 1995   
173                NaN  Threatened Species Protection Act 1995   
522                NaN  Threatened Species Protection Act 1995   
678                NaN  Threatened Species Protection Act 1995   
523                NaN  Threatened Species Protection Act 1995   

                                       new_description  \
216  Listed as Threatened - refer to https://www.th...   
300  Listed a

## Updates

In [40]:
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username,description
updates = pd.DataFrame(mergedstatuses[mergedstatuses['status_id'].notnull()])
updates.sort_values('scientificName')
updates['action'] = 'UPDATE'
#updates.loc[:,'action'] = 'UPDATE'
updates = updates[['action','scientificName','status_id','taxon_id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
updates.columns = updates.columns.str.replace("new_", "", regex=True)
updates = updates.rename(columns={'scientificName':'taxon_name',
                                  'status_id':'id'})
updates

action                 taxon_name      id taxon_id      status  \
216  UPDATE           Acacia axillaris  160726   775192          VU   
604  UPDATE  Alternanthera denticulata  160727   369665  Endangered   
605  UPDATE    Amphibromus macrorhinus  160728  1121410  Endangered   
243  UPDATE      Anogramma leptophylla  162081   180646  Vulnerable   
360  UPDATE           Aphelia gracilis  163877  1143399          NT   
..      ...                        ...     ...      ...         ...   
666  UPDATE     Thelymitra benthamiana  161102   520289  Endangered   
667  UPDATE       Thelymitra bracteata  161103   960004  Endangered   
671  UPDATE         Thelymitra malvina  161104   406524  Endangered   
674  UPDATE         Veronica notabilis  161105   791936  Endangered   
675  UPDATE           Viminaria juncea  161106   353911  Endangered   

    iucn_equivalent                               authority  \
216      Vulnerable  Threatened Species Protection Act 1995   
604      Vulnerable  Threatened Species Protection Act 1995   
605      Vulnerable  Threatened Species Protection Act 1995   
243      Vulnerable  Threatened Species Protection Act 1995   
360      Vulnerable  Threatened Species Protection Act 1995   
..              ...                                     ...   
666      Vulnerable  Threatened Species Protection Act 1995   
667      Vulnerable  Threatened Species Protection Act 1995   
671      Vulnerable  Threatened Species Protection Act 1995   
674      Vulnerable  Threatened Species Protection Act 1995   
675      Vulnerable  Threatened Species Protection Act 1995   

                                                   url geoprivacy place_id  \
216   https://id.biodiversity.org.au/node/apni/2911902   obscured     6829   
604  https://id.biodiversity.org.au/taxon/apni/5126...   obscured     6829   
605   https://id.biodiversity.org.au/node/apni/2916650   obscured     6829   
243  https://id.biodiversity.org.au/taxon/apni/5129...   obscured     6829   
360   https://id.biodiversity.org.au/node/apni/2899707   obscured     6829   
..                                                 ...        ...      ...   
666  https://id.biodiversity.org.au/taxon/apni/5141...   obscured     6829   
667  https://id.biodiversity.org.au/taxon/apni/5141...   obscured     6829   
671  https://id.biodiversity.org.au/taxon/apni/5141...   obscured     6829   
674   https://id.biodiversity.org.au/node/apni/2890187   obscured     6829   
675   https://id.biodiversity.org.au/node/apni/2904505   obscured     6829   

      username                                        description  
216  peggydnew  Listed as Threatened - refer to https://www.th...  
604  peggydnew  Listed as Threatened - refer to https://www.th...  
605  peggydnew  Listed as Threatened - refer to https://www.th...  
243  peggydnew  Listed as Threatened - refer to https://www.th...  
360  peggydnew  Listed as Threatened - refer to https://www.th...  
..         ...                                                ...  
666  peggydnew  Listed as Threatened - refer to https://www.th...  
667  peggydnew  Listed as Threatened - refer to https://www.th...  
671  peggydnew  Listed as Threatened - refer to https://www.th...  
674  peggydnew  Listed as Threatened - refer to https://www.th...  
675  peggydnew  Listed as Threatened - refer to https://www.th...  

[67 rows x 12 columns]

## No status in iNaturalist via straight scientificName match
The ACT records that didn't match up to a status in iNaturalist

In [41]:
# to add: those that have no inaturalist status - 532!!
noinatstatus = mergedstatuses[mergedstatuses['status_id'].isnull()]
# try to match the taxon name to something in inaturalist
noinatstatus = noinatstatus.merge(inattaxa, how="left", left_on="scientificName",right_on="scientificName")
noinatstatus

name bionet_geoprivacy  \
0                    Acacia pataczekii              open   
1                  Acacia siculiformis              open   
2                    Acacia ulicifolia              open   
3                     Acacia uncifolia              open   
4         Acanthiza pusilla archibaldi              open   
..                                 ...               ...   
612               Xerochrysum palustre              open   
613                   Zearaja maugeana              open   
614                  Zieria littoralis              open   
615  Zieria veronicea subsp. veronicea              open   
616           Zygophyllum billardierei              open   

                                                  lsid status_id  \
0     https://id.biodiversity.org.au/node/apni/2913609       NaN   
1     https://id.biodiversity.org.au/node/apni/2888960       NaN   
2     https://id.biodiversity.org.au/node/apni/2907350       NaN   
3     https://id.biodiversity.org.au/node/apni/2916270       NaN   
4                                        ALA_DR654_159       NaN   
..                                                 ...       ...   
612   https://id.biodiversity.org.au/node/apni/2890165       NaN   
613  https://biodiversity.org.au/afd/taxa/0d9d194f-...       NaN   
614   https://id.biodiversity.org.au/node/apni/2888662       NaN   
615   https://id.biodiversity.org.au/node/apni/2911634       NaN   
616  https://id.biodiversity.org.au/taxon/apni/5131...       NaN   

    scientificName taxon_id user_id description iucn authority  ... phylum  \
0              NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
1              NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
2              NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
3              NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
4              NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
..             ...      ...     ...         ...  ...       ...  ...    ...   
612            NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
613            NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
614            NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
615            NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
616            NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   

    class order family genus specificEpithet infraspecificEpithet modified  \
0     NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
1     NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
2     NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
3     NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
4     NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
..    ...   ...    ...   ...             ...                  ...      ...   
612   NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
613   NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
614   NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
615   NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
616   NaN   NaN    NaN   NaN             NaN                  NaN      NaN   

    taxonRank references  
0         NaN        NaN  
1         NaN        NaN  
2         NaN        NaN  
3         NaN        NaN  
4         NaN        NaN  
..        ...        ...  
612       NaN        NaN  
613       NaN        NaN  
614       NaN        NaN  
615       NaN        NaN  
616       NaN        NaN  

[617 rows x 37 columns]

In [42]:
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions

Empty DataFrame
Columns: [name, bionet_geoprivacy, lsid, status_id, scientificName, taxon_id, user_id, description, iucn, authority, status, geoprivacy, place_id, place_display_name, new_authority, new_description, new_url, new_geoprivacy, new_place_id, new_username, new_iucn_equivalent, new_status, id, taxonID, identifier, parentNameUsageID, kingdom, phylum, class, order, family, genus, specificEpithet, infraspecificEpithet, modified, taxonRank, references]
Index: []

[0 rows x 37 columns]

In [43]:
# there's no status but there is a matching inat taxon (id is the taxon id)
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions.sort_values(['scientificName'])
additions['action'] = 'ADD'
additions = additions[['action','scientificName','status_id','id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
additions.columns = additions.columns.str.replace("new_", "", regex=True)
additions = additions.rename(columns={'scientificName':'taxon_name',
                                      'id':'taxon_id',
                                  'status_id':'id'})
additions

Empty DataFrame
Columns: [action, taxon_name, id, taxon_id, status, iucn_equivalent, authority, url, geoprivacy, place_id, username, description]
Index: []

In [44]:
# write these to the file
pd.concat([updates,additions]).to_csv(sourcedir + "tas.csv", index=False)

In [45]:
# what didnt match to a taxon?
unknownToInat = noinatstatus[noinatstatus['id'].isna()]
unknownToInat

name bionet_geoprivacy  \
0                    Acacia pataczekii              open   
1                  Acacia siculiformis              open   
2                    Acacia ulicifolia              open   
3                     Acacia uncifolia              open   
4         Acanthiza pusilla archibaldi              open   
..                                 ...               ...   
612               Xerochrysum palustre              open   
613                   Zearaja maugeana              open   
614                  Zieria littoralis              open   
615  Zieria veronicea subsp. veronicea              open   
616           Zygophyllum billardierei              open   

                                                  lsid status_id  \
0     https://id.biodiversity.org.au/node/apni/2913609       NaN   
1     https://id.biodiversity.org.au/node/apni/2888960       NaN   
2     https://id.biodiversity.org.au/node/apni/2907350       NaN   
3     https://id.biodiversity.org.au/node/apni/2916270       NaN   
4                                        ALA_DR654_159       NaN   
..                                                 ...       ...   
612   https://id.biodiversity.org.au/node/apni/2890165       NaN   
613  https://biodiversity.org.au/afd/taxa/0d9d194f-...       NaN   
614   https://id.biodiversity.org.au/node/apni/2888662       NaN   
615   https://id.biodiversity.org.au/node/apni/2911634       NaN   
616  https://id.biodiversity.org.au/taxon/apni/5131...       NaN   

    scientificName taxon_id user_id description iucn authority  ... phylum  \
0              NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
1              NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
2              NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
3              NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
4              NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
..             ...      ...     ...         ...  ...       ...  ...    ...   
612            NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
613            NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
614            NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
615            NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   
616            NaN      NaN     NaN         NaN  NaN       NaN  ...    NaN   

    class order family genus specificEpithet infraspecificEpithet modified  \
0     NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
1     NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
2     NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
3     NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
4     NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
..    ...   ...    ...   ...             ...                  ...      ...   
612   NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
613   NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
614   NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
615   NaN   NaN    NaN   NaN             NaN                  NaN      NaN   
616   NaN   NaN    NaN   NaN             NaN                  NaN      NaN   

    taxonRank references  
0         NaN        NaN  
1         NaN        NaN  
2         NaN        NaN  
3         NaN        NaN  
4         NaN        NaN  
..        ...        ...  
612       NaN        NaN  
613       NaN        NaN  
614       NaN        NaN  
615       NaN        NaN  
616       NaN        NaN  

[617 rows x 37 columns]

### are there any that need to be removed?
TAS sensitive list count: 1
TAS conservation list count: 683
TAS inat statuses count: 0

updates to inat status: 67
additional inat status: 0
TAS statuses we can't find a taxon match for in iNaturalist: 617
total: (explainable via the various genus/section entries that we matched to in the taxonomy)

inat statuses left over:  that may need checking

In [46]:
# Stats
numsensitive = len(sensitivelist.index)
numconservation = len(conservationlist.index)
numupdates  = len(updates.index)
numadditions  = len(additions.index)
numnoinatstatus = len(noinatstatus.index)
numunknownToInat = len(unknownToInat.index)
d = {'Sensitive': [numsensitive],
    'Conservation': [numconservation],
    'Updates': [numupdates],
    'Additions': [numadditions],
    'No Inat Status': [numnoinatstatus],
    'Unknown to Inat': [numunknownToInat]}
statsdf = pd.DataFrame(data=d)
statsdf

Sensitive  Conservation  Updates  Additions  No Inat Status  \
0          1           683       67          0             617   

   Unknown to Inat  
0              617

In [47]:
# inat statuses that aren't in added or updated
inatstatuses[~inatstatuses['taxon_id'].isin(updates['taxon_id'])]


status_id taxon_id user_id place_id source_id  \
2557    160920  1038976  717122     6829             
2566    160929  1092444  717122     6829             
3353    267438  1413424  717122     6829             
1251    152435    19250  708886     9994     16652   
2599    161108   323791  717122     6829             
904     160736   353835  717122     6829             
1249    152433    38633  708886     9994     16652   
932     160737   401378  717122     6829             
959     160741   405825  717122     6829             
1252    152436    40743  708886     9994     16652   
1248    152432    41326  708886     9994     16652   
2588    160758   446738  717122     6829             
899     160766   699302  717122     6829             
902     160733   700972  717122     6829             
1247    152431    73180  708886     9994     16652   
1250    152434    74425  708886     9994     16652   
2598    161107   790432  717122     6829             
2526    160892   859959  717122     6829             
2586    160759   958863  717122     6829             

                                   authority            status  \
2557  Threatened Species Protection Act 1995        Endangered   
2566  Threatened Species Protection Act 1995        Endangered   
3353  Threatened Species Protection Act 1995        Endangered   
1251                               NT NRETAS        endangered   
2599  Threatened Species Protection Act 1995        Endangered   
904   Threatened Species Protection Act 1995        Endangered   
1249                               NT NRETAS        endangered   
932   Threatened Species Protection Act 1995        Endangered   
959   Threatened Species Protection Act 1995  Presumed extinct   
1252                               NT NRETAS        endangered   
1248                               NT NRETAS        endangered   
2588  Threatened Species Protection Act 1995        Endangered   
899   Threatened Species Protection Act 1995        Endangered   
902   Threatened Species Protection Act 1995           Extinct   
1247                               NT NRETAS        endangered   
1250                               NT NRETAS        endangered   
2598  Threatened Species Protection Act 1995        Endangered   
2526  Threatened Species Protection Act 1995        Endangered   
2586  Threatened Species Protection Act 1995        Endangered   

                                                    url description  \
2557  https://dpipwe.tas.gov.au/Documents/Pimelea-ax...               
2566  https://dpipwe.tas.gov.au/Documents/Spyridium-...               
3353  https://dpipwe.tas.gov.au/Documents/Pneumatopt...               
1251  https://lists.ala.org.au/speciesListItem/list/...               
2599  https://dpipwe.tas.gov.au/Documents/Zieria-ver...               
904   https://dpipwe.tas.gov.au/Documents/Blechnum-n...               
1249  https://lists.ala.org.au/speciesListItem/list/...               
932   https://dpipwe.tas.gov.au/Documents/Blechnum%2...               
959   https://dpipwe.tas.gov.au/Documents/Botrychium...               
1252  https://lists.ala.org.au/speciesListItem/list/...               
1248  https://lists.ala.org.au/speciesListItem/list/...               
2588  https://dpipwe.tas.gov.au/Documents/Corunastyl...               
899   https://dpipwe.tas.gov.au/Documents/Cyathea-Xm...               
902   https://dpipwe.tas.gov.au/Documents/Banksia-in...               
1247  https://lists.ala.org.au/speciesListItem/list/...               
1250  https://lists.ala.org.au/speciesListItem/list/...               
2598  https://dpipwe.tas.gov.au/Documents/Wurmbea-la...               
2526  https://dpipwe.tas.gov.au/Documents/Euphrasia-...               
2586  https://dpipwe.tas.gov.au/Documents/Corunastyl...               

     geoprivacy  ...         genus specificEpithet infraspecificEpithet  \
2557   obscured  ...       Pimelea        axiflora             axiflora   
2566   obscured  ...     Spyridi